### PR TITLE
DepositInput: Proof of possession, withdrawal credentials

### PR DESCRIPTION
**Pull request by prestonvanloon**
Part of #1510.

This adds proof of possession and withdrawal credentials in the shared keystore. 
This is in the shared keystore because I will need it in some tooling and it seems to make sense here.

~Question for reviewers: This uses protobuf wire protocol for the signature of the deposit data. Is it supposed to be SSZ or something else?~

Now using ssz encoding!

### CODE REVIEW

File: shared/keystore/deposit_input.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// signed by the deposit key.
+func DepositInput(depositKey *Key, withdrawalKey *Key) *pb.DepositInput {
+	di := &pb.DepositInput{
+		Pubkey:                      depositKey.PublicKey.Marshal(),
+		WithdrawalCredentialsHash32: withdrawalCredentialsHash(withdrawalKey),
+	}
+
+	// #nosec G104 -- This can only throw if di is nil so lets ignore the error.
+	b, _ := proto.Marshal(di)
+	di.ProofOfPossession = depositKey.SecretKey.Sign(b, params.BeaconConfig().DomainDeposit).Marshal()
```
Comment (rauljordan) [inline shared/keystore/deposit_input.go]: @prestonvanloon 

### Submit deposit

In phase 0, all incoming validator deposits originate from the Ethereum 1.0 PoW chain. Deposits are made to the [deposit contract](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#ethereum-10-deposit-contract) located at `DEPOSIT_CONTRACT_ADDRESS`.

To submit a deposit:

* Pack the validator's [initialization parameters](#initialization) into `deposit_input`, a [`DepositInput`](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#depositinput) SSZ object.
* Set `deposit_input.proof_of_possession = EMPTY_SIGNATURE`.
* Let `proof_of_possession` be the result of `bls_sign` of the `hash_tree_root(deposit_input)` with `domain=DOMAIN_DEPOSIT`.
* Set `deposit_input.proof_of_possession = proof_of_possession`.
* Let `amount` be the amount in Gwei to be deposited by the validator where `MIN_DEPOSIT_AMOUNT <= amount <= MAX_DEPOSIT_AMOUNT`.
* Send a transaction on the Ethereum 1.0 chain to `DEPOSIT_CONTRACT_ADDRESS` executing `deposit` along with `serialize(deposit_input)` as the singular `bytes` input along with a deposit `amount` in Gwei.

_Note_: Deposits made for the same `pubkey` are treated as for the same validator. A singular `Validator` will be added to `state.validator_registry` with each additional deposit amount added to the validator's balance. A validator can only be activated when total deposits for the validator pubkey meet or exceed `MAX_DEPOSIT_AMOUNT`.

Comment (prestonvanloon) [reply shared/keystore/deposit_input.go]: Thanks! Was this from the validator spec? I was looking at the main spec and didn't see any of this

Comment (rauljordan) [reply shared/keystore/deposit_input.go]: Yep from the latest honest validator spec

**Comment by rauljordan** [comment]
Looks great!

**Pull request merged**